Sparse Compressed mm avoid creating temp sparse

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104062

When mm forwards to addmm it creates a zeroed out self this tensor
should take options from the result not one of the sparse arguments.

The bug was leading to an error when calling linear with an `out` kwarg.

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

amjames committed:
Sparse Compressed mm avoid creating temp sparse

When mm forwards to addmm it creates a zeroed out self this tensor
should take options from the result not one of the sparse arguments.

The bug was leading to an error when calling linear with an `out` kwarg.

[ghstack-poisoned]

nikitaved reviewed (approved):
👍 LGTM! Thank you, @amjames !

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -735,7 +735,7 @@ Tensor& _sparse_csr_mm_out(
     const Tensor& mat2,
     Tensor& result) {
-  auto zero = at::zeros({mat1.size(0), mat2.size(1)}, mat2.options());
+  auto zero = at::zeros({mat1.size(0), mat2.size(1)}, result.options());
   return at::addmm_out(result, zero, mat1, mat2, 0.0, 1.0);
```
It will not break if `result` is sparse, right?

amjames replied:
If result is block-sparse, yes it will fail under some conditions. The primary issue is when calling zero forwards to `empty` getting a block-sparse layout it chokes out without any information about what block size to use.

Actually using `zeros_like(result)` should be sufficient here. If we do that, we at least do not get a transitive error from empty. Instead the error would come from the combination of layouts not being supported for `A + B@C`.

amjames committed:
Update on "Sparse Compressed mm avoid creating temp sparse"


When mm forwards to addmm it creates a zeroed out self this tensor
should take options from the result not one of the sparse arguments.

The bug was leading to an error when calling linear with an `out` kwarg.

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

amjames committed:
Update on "Sparse Compressed mm avoid creating temp sparse"


When mm forwards to addmm it creates a zeroed out self this tensor
should take options from the result not one of the sparse arguments.

The bug was leading to an error when calling linear with an `out` kwarg.

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

pearu reviewed (approved):
LGTM! Thanks, @amjames!

amjames commented:
@pytorchbot merge
